### DISCUSSION TITLE
Migration to PocketBase, advice needed

### DISCUSSION THREAD

**Discussion by ttijl** [pocketbase/pocketbase#2539] [Q&A] [answered]
Hello i have a few questions.
I currently use mongoDB for a few simple personal projects. But now i want real time data and better security, pocket base seems amazing for me, since i already host everything myself.
Few questions:
 - Is there a way to make a custom authentication system with custom logic?
 - Is there any way to group collections, to keep different projects apart or is this planned?

**Comment by ganigeorgiev** [comment, marked as answer]
> Is there a way to make a custom authentication system with custom logic?

At the moment you can extend PocketBase only with Go - https://pocketbase.io/docs/use-as-framework/.
Support for JS app hooks via [goja](https://github.com/dop251/goja) is planned for the next v0.17.0 release but there are no fixed ETAs.

> Is there any way to group collections, to keep different projects apart or is this planned?

There is no dedicated mechanism to group different projects and at least for v1 there are no plans to add support for this.
The recommended approach is to have one PocketBase instance per project (this way it will be also easier to migrate one of the projects on different server in case a single shared server is no longer sufficient). Alternatively, if you really want to keep multiple projects as part of a single PocketBase instance, you can add a unique project prefix to your collections, but keep in mind that this may complicate the [API rules management](https://pocketbase.io/docs/api-rules-and-filters/).

**Reply by ttijl** [reply]
Thanks for your fast response.
> At the moment you can extend PocketBase only with Go - https://pocketbase.io/docs/use-as-framework/.
Support for JS app hooks via [goja](https://github.com/dop251/goja) is planned for the next v0.17.0 release but there are no fixed ETAs.

Will look into writing my own Go.. JS Hooks also would also be cool. 

> The recommended approach is to have one PocketBase instance per project (this way it will be also easier to migrate one of the project on different server in case a single shared server is no longer sufficient).

For me that's probably not a option because i have many tiny projects... Will try using a collection prefix.

I will probably only migrate a part of my Projects to a PocketBase instance.
Thanks for the amazing work, and cool that something like this is open source and easy to self host! And I'm exited about what PocketBase can become in the future!

**Comment by gardc** [comment]
I’m also migrating to pocketbase from a MongoDB based project. Is there a recommended way to migrate the MongoDB data into the SQLite pocketbase db?

Currently I’m experimenting with writing a script that is manually iterating every mdb document, parsing it and inserting it into a local SQLite DB, and finally inserting all that data into the pocketbase DB. Does that sound reasonable, or does anyone have experience with a simpler way of doing this?

**Reply by ganigeorgiev** [reply]
You can check older discussions like https://github.com/pocketbase/pocketbase/discussions/2014, https://github.com/pocketbase/pocketbase/discussions/2088, etc.
Note that if you go with the SQLite import script route you have to define the schemas in the `_collections` table (_or create them manually beforehand from the UI; on collection save it will create its underlying records table_).

**Comment by pnmcosta** [comment]
> * Is there a way to make a custom authentication system with custom logic?

Do you mean use a custom [oauth](https://pocketbase.io/docs/authentication/#oauth2-integration) provider? 

I have done so on my project that leverages PB as a framework, in 3 distinct ways:

1. Shopify integration, I have custom routes to handle it's oauth and session token exchange, completely bypassed PB's oauth, but leverage a lot of the existing `apis.*` and am still able use the js sdk. I store integration details and access tokens in a separate collection that will be encrypted like OTB settings do.
2. Facebook integration, I use the existing oauth provider on PB to do Facebook Business Login, works great, I do handle `OnRecordAfterAuthWithOAuth2Request` and `OnRecordAfterUnlinkExternalAuthRequest` to also manage the integration details and access tokens in the separate collection.
3. Instagram integration, I could have used the exiting `oidc` providers, unfortunately IG does not supply user details in openid standard so had to [contribute](https://github.com/pocketbase/pocketbase/pull/2534) with it's own provider, hopefully it will be approved 🤞 
 
However as long as the user fetch details match [this](https://github.com/pocketbase/pocketbase/blob/master/tools/auth/oidc.go#L46), you should be able to use the existing oidc providers and leverage the Record OAuth [hooks](https://pocketbase.io/docs/event-hooks/#record-auth-api-hooks) for most of custom logic.

Let me know if you have any questions, happy to help the community.
